src/benchmark.cc

```diff
@@ -281,26 +343,75 @@ void RunBenchmarks(const std::vector<BenchmarkInstance>& benchmarks,
     flushStreams(display_reporter);
     flushStreams(file_reporter);
 
-    for (const auto& benchmark : benchmarks) {
-      RunResults run_results = RunBenchmark(benchmark, &complexity_reports);
+    // Without random interleaving, benchmarks are executed in the order of:
+    //   A, A, ..., A, B, B, ..., B, C, C, ..., C, ...
+    // That is, repetition is within RunBenchmark(), hence the name
+    // inner_repetitions.
+    // With random interleaving, benchmarks are executed in the order of:
+    //  {Random order of A, B, C, ...}, {Random order of A, B, C, ...}, ...
+    // That is, repetitions is outside of RunBenchmark(), hence the name
+    // outer_repetitions.
+    int inner_repetitions =
+        FLAGS_benchmark_enable_random_interleaving ? 1 : GetRepetitions();
+    int outer_repetitions =
+        FLAGS_benchmark_enable_random_interleaving ? GetRepetitions() : 1;
+    std::vector<size_t> benchmark_indices(benchmarks.size());
+    for (size_t i = 0; i < benchmarks.size(); ++i) {
+      benchmark_indices[i] = i;
+    }
 
-      auto report = [&run_results](BenchmarkReporter* reporter,
-                                   bool report_aggregates_only) {
+    auto report = [flushStreams, display_reporter, file_reporter](
+        const RunResults& run_results) {
+      auto report_one = [](BenchmarkReporter* reporter,
+                           bool aggregates_only,
+                           const RunResults& results) {
         assert(reporter);
         // If there are no aggregates, do output non-aggregates.
-        report_aggregates_only &= !run_results.aggregates_only.empty();
-        if (!report_aggregates_only)
-          reporter->ReportRuns(run_results.non_aggregates);
-        if (!run_results.aggregates_only.empty())
-          reporter->ReportRuns(run_results.aggregates_only);
+        aggregates_only &= !results.aggregates_only.empty();
+        if (!aggregates_only)
+          reporter->ReportRuns(results.non_aggregates);
+        if (!results.aggregates_only.empty())
+          reporter->ReportRuns(results.aggregates_only);
       };
 
-      report(display_reporter, run_results.display_report_aggregates_only);
+      report_one(display_reporter, run_results.display_report_aggregates_only,
+                 run_results);
       if (file_reporter)
-        report(file_reporter, run_results.file_report_aggregates_only);
+        report_one(file_reporter, run_results.file_report_aggregates_only,
+                   run_results);
 
       flushStreams(display_reporter);
       flushStreams(file_reporter);
+    };
```

Comment:
this looks like it should be a separate (private.. static?) method now.

Reply:
Done